app/models/proposal.rb

```diff
@@ -93,36 +99,38 @@ def users
     results.compact
   end
 
-  # Set the approver list, from any start state
-  # This overrides the `through` relation but provides parity to the accessor
-  def approvers=(approver_list)
-    approvals = approver_list.each_with_index.map do |approver, idx|
-      approval = self.existing_approval_for(approver)
-      approval ||= Approvals::Individual.new(user: approver, proposal: self)
-      approval.position = idx + 1   # start with 1
-      approval
+  def root_approval=(root)
+    old_approvals = self.approvals.to_a
+
+    approval_list = root.preorder_list
+    self.approvals = approval_list
+    # position may be out of whack, so we reset it
+    approval_list.each_with_index do |approval, idx|
+      approval.set_list_position(idx + 1)   # start with 1
+    end
+
+    old_approvals.each do |old|
+      unless approval_list.include?(old)
+        old.destroy()
+      end
     end
-    self.approvals = approvals
-    self.kickstart_approvals()
-    self.reload   # include the changes in kickstart_approvals
+
+    root.initialize!
     self.reset_status()
   end
 
-  # Trigger the appropriate approval, from any start state
-  def kickstart_approvals()
-    actionable = self.approvals.actionable
-    pending = self.approvals.pending
-    if self.parallel?
-      pending.update_all(status: 'actionable')
-    elsif self.linear? && actionable.empty? && pending.any?
-      pending.first.initialize!
+  # convenience wrapper for setting a single approver
+  def approver=(approver)
+    # Don't recreate the approval
+    existing = self.existing_approval_for(approver)
+    if existing.nil?
+      self.root_approval = Approvals::Individual.new(user: approver)
```

Comment:
Oh nevermind, looks like that might be taken care of in `#root_approval=`...? Anyway, we should test for it.